---
title: "R Notebook"
output:
  html_document:
    df_print: paged
params:
  config: ''
---

```{r libraries}
library(Seurat,quietly = TRUE,warn.conflicts = FALSE)
library(gridExtra,quietly = TRUE,warn.conflicts = FALSE)
library(dplyr,quietly = TRUE,warn.conflicts = FALSE)
library(Signac,quietly = TRUE,warn.conflicts = FALSE)
library(EnsDb.Mmusculus.v79,quietly = TRUE,warn.conflicts = FALSE)
library(ggplot2,quietly = TRUE,warn.conflicts = FALSE)



set.seed(100)


```


```{r set_root_dir}

setwd('../../')
knitr::opts_knit$set(root.dir = getwd())
```
# Read config

```{r read_config}
config <- yaml::read_yaml(params$config)
print(config)
```

# load annotations

```{r load_gene_annotations}
gene.coords <- ensembldb::genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "protein_coding")
lncRNA.coords <- ensembldb::genes(EnsDb.Mmusculus.v79, filter = ~ gene_biotype == "lincRNA")
gene.coords <- c(gene.coords,lncRNA.coords)

seqlevelsStyle(gene.coords) <- 'UCSC'
genebody.coords <- keepStandardChromosomes(gene.coords, pruning.mode = 'coarse')

# Flatten the overlapping genes and extend by 2kb upstream of promoters
genebody.coords.flat <- GenomicRanges::reduce(x = genebody.coords)
genebodyandpromoter.coords.flat <- Signac::Extend(genebody.coords.flat,upstream = 2000)

# Retrieve gene names from the original annotation (lost because of flatenning)
genebodyandpromoter.coords.flat$name<- gene.coords[nearest(genebodyandpromoter.coords.flat,genebody.coords)]$gene_name
```

```{r load_and_merge}
assay = "peaksMB"
brain.ls <- lapply(names(config$input$H3K4me3),function(x){
  
  if(any(!c("seurat_object","fragments","peaks" ) %in% names(config$input$H3K4me3[[x]]))){
    stop("Need to specify following in the config yaml file [see example]: seurat_object, fragments, peaks")
  }
  
  if(any(!file.exists(unlist(config$input$H3K27me3[[x]])))){
    stop("Need to specify following in the config yaml file [see example]: seurat_object, fragments, peaks")
  }
  
  # Read the seurat object
  brain                     <- readRDS(config$input$H3K4me3[[x]]$seurat_object)
  DefaultAssay(brain)       <- assay
  
  brain$age <- unlist(strsplit(names(config$input$H3K4me3[x]),"_"))[1]
  brain$GFP <- unlist(strsplit(names(config$input$H3K4me3[x]),"_"))[2]
  
  brain
})

brain.merged <- Reduce(function(x,y) merge(x,y), brain.ls)
```

```{r filter_QC}
 # Filter blacklisted regions / cells
blacklist.overlaps <- subsetByOverlaps(x = StringToGRanges(rownames(brain.merged)),ranges =   blacklist_mm10,type = 'any')
blacklist.bins     <- which(rownames(brain.merged) %in% GRangesToString(blacklist.overlaps))
blacklist.cells    <- colnames(brain.merged[,brain.merged$blacklist_region_fragments > 5])
blacklist.cells    <- which(colnames(brain.merged) %in% blacklist.cells)

mat.bins <- brain.merged[['peaksMB']]@counts
mat.bins <- mat.bins[-blacklist.bins,-blacklist.cells]

mat.genes <- brain.merged[['GA']]@counts
mat.genes <- mat.genes[,colnames(mat.bins)]

meta.data <- brain.merged@meta.data
meta.data <- meta.data[colnames(mat.bins),]

brain.merged <- CreateSeuratObject(counts = mat.bins,project = 'scCT',assay = 'bins_5kb',meta.data = meta.data)
brain.merged[['GA']] <- CreateAssayObject(counts = mat.genes)

brain.merged  <- brain.merged[,brain.merged$logUMI > 2]

DefaultAssay(brain.merged) <- 'bins_5kb'

```

```{r cluster_and_dimreduce}
# Re-do the dimreduction
brain.merged <- RunTFIDF(brain.merged)
brain.merged <- FindTopFeatures(brain.merged, min.cutoff = 'q0')
  
brain.merged <- RunSVD(
  object = brain.merged,
  assay = 'bins_5kb',
  reduction.key = 'LSI_',
  reduction.name = 'lsi',
  n = 50
)
  
brain.merged <- RunUMAP(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:40
)

brain.merged <- FindNeighbors(
  object = brain.merged,
  reduction = 'lsi',
  dims = 2:50
)
brain.merged <- FindClusters(
  object = brain.merged,
  algorithm = 3,
  resolution = 0.8,
  verbose = FALSE
)

DimPlot(brain.merged,label=TRUE)

```


```{r plots}

p1 <- DimPlot(brain.merged,group.by='sample',label=TRUE) + NoLegend()
p2 <- DimPlot(brain.merged,label=TRUE)+ NoLegend()

p6 <- DimPlot(brain.merged,label=TRUE,group.by = 'age')+ NoLegend()
p7 <- DimPlot(brain.merged,label=TRUE,group.by = 'GFP')+ NoLegend()

p3 <- DimPlot(brain.merged[,brain.merged$sample == 'H3K4me3_P15_GFP+_N1'],label=TRUE)+ NoLegend()
p4 <- DimPlot(brain.merged[,brain.merged$sample == 'H3K4me3_P15_GFP-_N1'],label=TRUE)+ NoLegend()
p5 <- DimPlot(brain.merged[,brain.merged$sample %in% c("H3K4me3_N1","H3K4me3_N2")],label=TRUE)+ NoLegend()

p1 + p2
p3 + p4
p6 + p7
p5
FeaturePlot(brain.merged,'logUMI',max.cutoff = 3) + scale_color_viridis_c()
```






```{r load_RNA_markers}
markers.Sox10 <- read.csv2(file='results/Sox10_RNA/clustering/GFP/markers.csv',row.names = 1)
markers <- read.csv2(file='results/Sten_RNA/clustering/sten_RNA_markers_temp.csv',row.names = 1)
```





```{r Plot_RNA_modules_projection,fig.width=6,fig.height=6}
assay='GA'

markers.Sox10$cluster <- paste0('Sox10_',markers.Sox10$cluster)
markers               <- rbind(markers,markers.Sox10)

brain.markers.ls <- lapply(unique(markers$cluster),function(x){
  marker.genes <- markers[markers$cluster==x & markers$p_val_adj < 0.05 & markers$avg_logFC > 0,"gene"]
  marker.genes <- head(marker.genes,200)
  marker.genes.agg.exp <- Matrix::colSums(brain.merged[[assay]]@data[rownames(brain.merged[[assay]]) %in% marker.genes,])
  marker.genes.agg.exp <- marker.genes.agg.exp / Matrix::colSums(brain.merged[[assay]]@counts)
  return(marker.genes.agg.exp)
})

names(brain.markers.ls)  <- gsub("-",".",unique(markers$cluster))
markers.agg.df           <- do.call(cbind,brain.markers.ls)
colnames(markers.agg.df) <- paste0("marker_",gsub(" ",".",colnames(markers.agg.df)))


brain.merged <- AddMetaData(brain.merged,metadata = markers.agg.df,col.name = colnames(markers.agg.df))

# Plotting

p1 <- DimPlot(brain.merged,group.by='sample',label=TRUE) + NoLegend()
p2 <- DimPlot(brain.merged,label=TRUE)+ NoLegend()

p3 <- DimPlot(brain.merged[,brain.merged$sample == 'H3K4me3_P15_GFP+_N1'],label=TRUE)+ NoLegend()
p4 <-  DimPlot(brain.merged[,brain.merged$sample != 'H3K4me3_P15_GFP+_N1'],label=TRUE)+ NoLegend()

p1 + p2
p3 + p4
p2 + coord_cartesian(xlim=c(-5,5),ylim=c(-5,5))


lapply(as.character(colnames(markers.agg.df)),function(x){
  FeaturePlot(brain.merged,x,max.cutoff = quantile(markers.agg.df[,x],0.9,na.rm=TRUE),min.cutoff = quantile(markers.agg.df[,x],0.05,na.rm=TRUE)) + scale_color_viridis_c() + NoLegend()+ coord_cartesian(xlim=c(-8,8),ylim=c(-8,8))
  #FeaturePlot(brain,x,max.cutoff = 0.08,min.cutoff = 0.02) + scale_color_viridis_c(limits=c(0.02,0.08)) 
  })


```

############################################################# 



```{r Rename_the_clusters,fig.width=10,fig.height=10}
# Remove small clusters
small.clusters <- table(brain.merged@active.ident)
small.clusters <- names(small.clusters)[9:length(small.clusters)]
brain.merged <- brain.merged[,!brain.merged@active.ident %in% small.clusters]

# Rename the clusters
brain.merged <- RenameIdents(brain.merged, '0' = 'Astrocytes',
                                    '1' = 'mOL',
                                    '2' = 'OEC',
                                    '3' = 'Inh_neurons',
                                    '4' = 'VLMCs',
                                    '5' = 'Exc_neurons',
                                    '6' = 'OPC',
                                    '7' = 'Micoglia')

brain.merged$cell_type <- brain.merged@active.ident

p1 <- DimPlot(object = brain.merged, group.by='age',label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend()
p2 <- DimPlot(object = brain.merged, label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend()
p3 <- DimPlot(object = brain.merged, group.by='GFP',label = TRUE,reduction = 'umap',pt.size=0.1) + NoLegend() + scale_color_manual(values = RColorBrewer::brewer.pal(n = 9,name='Set1')[c(4,3)])

p1
p2
p3
```





```{r plots_2, fig.width=10,fig.height=10}

p1 <- DimPlot(brain.merged,group.by='sample',label=TRUE) + NoLegend()
p2 <- DimPlot(brain.merged,label=TRUE)+ NoLegend()

p6 <- DimPlot(brain.merged,label=TRUE,group.by = 'age')+ NoLegend()
p7 <- DimPlot(brain.merged,label=TRUE,group.by = 'GFP')+ NoLegend() + scale_color_manual(values = RColorBrewer::brewer.pal(n = 9,name='Set1')[c(4,3)])

p.ls <- lapply(names(table(brain.merged$sample)),function(x){
  DimPlot(brain.merged[,brain.merged$sample == x],label=TRUE)+ NoLegend() + ggtitle(x)
})


p1 + p2
p6 + p7
do.call(grid.arrange,p.ls)

FeaturePlot(brain.merged,'logUMI',max.cutoff = 3) + scale_color_viridis_c()
```




```{r find_markers}
brain.merged<- NormalizeData(brain.merged,normalization.method = 'LogNormalize',scale.factor=10000)
markers <- FindAllMarkers(brain.merged,min.pct = 0.01,logfc.threshold = 0.1,slot = 'data')

markers$closest_gene <- ClosestFeature(StringToGRanges(markers$gene ), genebodyandpromoter.coords.flat)$name
#View(markers)

markers.top <- markers %>% group_by(cluster) %>% top_n(n = 50,wt = avg_logFC)

saveRDS(brain.merged,"results/H3K4me3/clustering/01.clustering.Rds")
write.csv2(x = markers,file = "results/H3K4me3/clustering/markers.csv")
write.csv2(x = markers.top,file = "results/H3K4me3/clustering/markers_top.csv")

```

```{r export_bw}
fragments1 <- config$input$H3K4me3$`P15_GFP-`$fragments
fragments1 <- rtracklayer::import(fragments1,format = "bed")

fragments2 <- config$input$H3K4me3$`P15_GFP+`$fragments
fragments2 <- rtracklayer::import(fragments2,format = "bed")

fragments3 <- config$input$H3K4me3$`P25_GFP+`$fragments
fragments3 <- rtracklayer::import(fragments3,format = "bed")

fragments  <- c(fragments1,fragments2,fragments3)



chrom.sizes <- read.table(url('http://hgdownload.cse.ucsc.edu/goldenPath/mm10/bigZips/mm10.chrom.sizes'),sep="\t",stringsAsFactors = FALSE)
chrom.sizes <- chrom.sizes[1:21,]

exportBW <- function(object,cluster,fragments){
  if(class(object) == "Seurat"){
    cells <- rownames(object@meta.data[object@active.ident == cluster,])
  }
  
  new_read <- GRanges(seqnames = chrom.sizes[,1], 
          ranges =IRanges(start = as.numeric(chrom.sizes[,2]),
                          width=1),
          name = "in_silico_extra_read",
          score = 0
          )
  
  fragments.x <- fragments$name %in% cells
  fragments.x <- fragments[fragments.x]
  fragments.x <- c(fragments.x,new_read)
  
  
  coverage.x <- coverage(fragments.x)
  
  # Option A - normalize by number of reads per sample
  coverage.x <- coverage.x/(length(fragments.x)/1e6)
  
  # Option B - normalize by mean signal (~ enrichment of mean signal)
  # coverage.x <- coverage.x / mean(unlist(coverage.x))
  
  rtracklayer::export.bw(object = coverage.x,paste0("results/H3K4me3/clustering/bigwig/cluster_",cluster,".bw"))
}

lapply(levels(brain.merged@active.ident),function(x){
  exportBW(brain.merged,x,fragments)
})

fragments.x <- fragments[fragments$name %in% colnames(brain.merged)]
coverage.x  <- coverage(fragments.x) / length(fragments.x)
rtracklayer::export.bw(object = coverage.x,con = "results/H3K4me3/clustering/bigwig/clusters_all/clusters_all.bw")

```

# Export marker bed file 

```{r do_metagene}
clusters_order = levels(brain.merged@active.ident)

# Get top markers
top.markers <- markers %>% group_by(cluster) %>% dplyr::filter(avg_logFC > 0) %>% dplyr::filter(p_val_adj < 0.05) %>% top_n(500, wt=avg_logFC)

# Reorder markers
top.markers$cluster <- factor(as.character(top.markers$cluster),levels=clusters_order[1:sum(table(top.markers$cluster) > 0)])

top.markers         <- as.data.frame(top.markers)
top.markers         <- top.markers[order(top.markers$cluster),]


dir.create("results/H3K4me3/clustering/markers_bed/",recursive=TRUE)

lapply(levels(top.markers$cluster),function(x){
  top.markers.x <- top.markers[top.markers$cluster == x,]
  top.markers.x <- sort(StringToGRanges(top.markers.x$gene))
  rtracklayer::export.bed(object = top.markers.x,con = paste0("results/H3K4me3/clustering/markers_bed/markers_cluster_",x,".bed"))
})

bw.files <- list.files(path = "results/H3K4me3/clustering/bigwig/",pattern = "*.bw")


system(paste0("/usr/local/anaconda3/bin/computeMatrix reference-point   -S results/H3K4me3/clustering/bigwig/*.bw",  
                                                                      " -R results/H3K4me3/clustering/markers_bed/*.bed", 
                                                                      " -o results/H3K4me3/clustering/markers_bed/metagene_matrix.txt.gz",
                                                                      " -a 10000 ", 
                                                                      " -b 10000 ",
                                                                      " -p 4",sep= ' '))


 
 system(paste0("/usr/local/anaconda3/bin/plotHeatmap   -m results/H3K4me3/clustering/markers_bed/metagene_matrix.txt.gz", 
                                                     " -o results/H3K4me3/clustering/markers_bed/metagene_matrix.pdf", 
                                                     " --sortRegions no",
                                                     " --colorList Darkblue,white,red",
                                                     " --refPointLabel peak"))

```


```{r do_heatmap}

DoHeatmapMB <- function(object, markers,fraction = 0.2,slot = "data",assay="bins_5kb", clusterColors){
  
  object <- object[,sample(colnames(object),length(colnames(object)) * fraction)]
  
  row_annotation.df           <- data.frame(markers=markers$cluster)
  rownames(row_annotation.df) <- markers$gene
  col_annotation.df           <- data.frame(clusters=Idents(object))
  col_annotation.df$sample    <- object$sample
  col_annotation.df           <- col_annotation.df[order(col_annotation.df$clusters,col_annotation.df$sample),]
  
  heatmap.mat <- GetAssayData(object = object, slot,assay = assay)
  heatmap.mat <- heatmap.mat[rownames(row_annotation.df),rownames(col_annotation.df)]
  
  print(dim(heatmap.mat))
  
  pheatmap::pheatmap(heatmap.mat,
                     cluster_cols = FALSE, 
                     cluster_rows = FALSE,
                     annotation_row = row_annotation.df,
                     annotation_col = col_annotation.df,
                     color = viridis::viridis(256),
                     show_rownames = FALSE,
                     show_colnames = FALSE,
                     labels_col = "",labels_row = "",
                     annotation_colors = clusterColors)
}

ncells        <- dim(brain.merged)[2]
cells_to_plot <- colnames(brain.merged)[sample(1:ncells,ncells * 0.4)]
cells_to_plot <- tibble::rownames_to_column(brain.merged.merged[,cells_to_plot]@meta.data) %>% arrange(seurat_clusters,sample) %>% pull(rowname)
clusters_order <- levels(brain.merged@active.ident)

# Get top markers
top.markers <- markers %>% group_by(cluster) %>% dplyr::filter(avg_logFC > 0) %>% top_n(100, wt=avg_logFC)
top.markers <- top.markers[which(!top.markers$gene %in% names(table(top.markers$gene)[table(top.markers$gene) > 1])),] # remove non-unique

# Reorder markers
top.markers$cluster <- factor(as.character(top.markers$cluster),levels=clusters_order)
top.markers         <- top.markers[order(top.markers$cluster),]

# Reorder cells
Idents(brain.merged) <- factor(as.character(Idents(brain.merged)),levels=clusters_order)

# Create list of cluster colors
clusterColors <- rep(list(scales::hue_pal()(length(levels(brain.merged@active.ident)))),2)
names(clusterColors) <- c("markers","clusters")
  
names(clusterColors[[1]]) <- levels(brain.merged@active.ident)
names(clusterColors[[2]]) <- levels(brain.merged@active.ident)


png(filename = "results/H3K4me3/clustering/bins_heatmap.png",width=1000,height=1000)
DoHeatmapMB(brain.merged,
            markers = top.markers,
            clusterColors = clusterColors,
            slot = 'data')
dev.off()

DoHeatmapMB(brain.merged,
            markers = top.markers,
            clusterColors = clusterColors,
            slot = 'data')


```













