---
title: "R Notebook"
output: html_notebook
params:
  out_prefix: "~/mount/CT/git_test/results/"
---




```{r}
library(Seurat)
library(Signac)
library(ggplot2)
library(EnsDb.Mmusculus.v79)
library(dplyr)
library(ggthemes)
library(pheatmap)
library(parallel)
library(tictoc)
library(Hmisc)
library(slingshot)

set.seed(2020)

# order clusters
clusters_order <- c("Astrocytes","VLMC","mOL","Neurons_1","Neurons_2","Neurons_3","OEC","OPC","Microglia")

# Colors definition - consistent accross the paper
CTcolors        <- scales::hue_pal()(9)
names(CTcolors) <- clusters_order
```



#Spreading

```{r load_data}

# Seurat object
brain.h3k4me3 <- readRDS(paste0(params$out_prefix,"H3K4me3/clustering/01.clustering.Rds"))


# Fragments
fragments <- paste0(params$out_prefix,"merged/H3K4me3/fragments.tsv.gz")
fragments <- rtracklayer::import(fragments,format='bed')
fragments <- fragments[fragments$name %in% colnames(brain.h3k4me3)]

#fragments <- resize(fragments,fix='center',width = 1)

# scRNA-seq marker genes
RNA.markers          <- read.csv2(file=paste0(params$out_prefix,'Sox10_RNA/clustering/GFP/markers.csv'),row.names = 1)
RNA.markers.positive <- RNA.markers[RNA.markers$avg_logFC > 0 & RNA.markers$p_val_adj < 0.05,]

# Promoters information
promoters.strict                 <- GenomicFeatures::promoters(EnsDb.Mmusculus.v79,downstream = 1,upstream = 0)
seqlevelsStyle(promoters.strict) <- "UCSC"

# Add gene name information
genes.ann                        <- as.data.frame(genes(EnsDb.Mmusculus.v79))[,c('gene_id','gene_name')]
promoters.strict$gene_name       <- genes.ann[promoters.strict$gene_id,'gene_name']

# Load peaks for H3K4me3 identified with MACS
peaks                  <- rtracklayer::import(paste0(params$out_prefix,"merged/H3K4me3/macs/broad/H3K4me3_peaks.broadPeak"))
peaks                  <- keepStandardChromosomes(peaks,pruning.mode = 'coarse')

# Filter peaks for those that overlap promoters
peaks.op                  <- findOverlaps(peaks,Extend(promoters.strict,upstream = 2000,downstream=200))
peaks.promoters           <- peaks[queryHits(peaks.op)]
peaks.promoters$gene      <- promoters.strict$gene_name[subjectHits(peaks.op)]
strand(peaks.promoters)   <- strand(promoters.strict[subjectHits(peaks.op)])

peaks.promoters           <- unique(peaks.promoters)
peaks.promoters
# 18121 out of ~30 000 peaks

```



```{r}
# RNA markers for mOLs - top 100
RNA.markers.mOL        <- head(RNA.markers.positive[RNA.markers.positive$cluster=="Oligodendrocytes",],100)

# Filter promoters for marker promtoers
promoters.mOL          <- unique(promoters.strict[promoters.strict$gene_name %in% RNA.markers.mOL$gene])
promoters.mOL          <- promoters.mOL[promoters.mOL$tx_biotype %in% c("protein_coding","lincRNA")]

# Filter promoters for promoters with peaks - # 1052 promoters
promoters.mOL          <- promoters.mOL[unique(queryHits(findOverlaps(promoters.mOL,peaks)))]


# Extend promoters
promoters.mOL.extended <- Signac::Extend(promoters.mOL,upstream = 20000,downstream = 20000)


# filter peaks to get mOL-specific peaks 
peaks.mOL              <- unique(peaks.promoters[peaks.promoters$gene %in% RNA.markers.mOL$gene])
start(peaks.mOL[strand(peaks.mOL) == "+"]) <- end(peaks.mOL[strand(peaks.mOL) == "+"])
end(peaks.mOL[strand(peaks.mOL) == "-"]) <- start(peaks.mOL[strand(peaks.mOL) == "-"])




```


```{r,fig.width=2,fig.height=4}
DefaultAssay(brain.h3k4me3) <- "GA"

# Find mOL markers realtive to OPCs
markers.mOL <- FindMarkers(brain.h3k4me3,ident.1 = 'mOL',min.pct = 0.02,only.pos = TRUE)

# Filter markers for positive and siginificant - 182 markers
markers.mOL <- markers.mOL[markers.mOL$avg_logFC > 0 & markers.mOL$p_val_adj < 0.05,]
markers.mOL <- markers.mOL[rownames(markers.mOL) %in% rownames(brain.h3k4me3),]

# Fetch the data for the markers
a <- Matrix.utils::aggregate.Matrix(t(brain.h3k4me3[['GA']]@counts[rownames(markers.mOL),]),groupings = brain.h3k4me3@active.ident,fun = 'median')

# Plot GA for the markers for OLs and OPCs
ggplot(data=reshape2::melt(as.data.frame(t(a[c('mOL','OPC'),])))) + geom_boxplot(aes(x=variable,y=value),outlier.shape = NA) + coord_cartesian(ylim=c(0,600))

```





```{r}
###################################################
# mOL sigeature is defined from the scCT data

# Fetch the whole dataset count matrix and normalize
mat <- apply(brain.h3k4me3[['GA']]@counts,2,function(x){
  x <- x/sum(x)
  x
})

# Get summed mOL signature ratio
mOL.signature         <- Matrix::colMeans(mat[rownames(markers.mOL),])
mOL.signature         <- as.data.frame(mOL.signature)
mOL.signature$cluster <- brain.h3k4me3@active.ident[rownames(mOL.signature)]

# plot
ggplot(data=mOL.signature) + geom_boxplot(aes(x=cluster,y=mOL.signature,fill=cluster),outlier.shape = NA) + theme_few()

# Add to object Metadata
brain.h3k4me3 <- AddMetaData(brain.h3k4me3,mOL.signature[,'mOL.signature',drop=FALSE])

# Plot
FeaturePlot(brain.h3k4me3,'mOL.signature',min.cutoff=0.0001,max.cutoff = 0.00025,pt.size=0.02) + scale_color_viridis_c()
```


```{r fragments_prefilter}
# mOL promoters are defined from the scRNA-seq datasets as promoters of genes specifically expressed in mOLs
# Filter fragments for only fragments within the target promoters (extended by 20 kb on each side)
op                      <- queryHits(findOverlaps(fragments,promoters.mOL.extended))
fragments.promoters.mOL <- fragments[op]

tic()
fragments.ls <- mclapply(colnames(brain.h3k4me3),mc.cores = 4,FUN = function(x){
  fragments[fragments$name %in% x]
})
toc()

names(fragments.ls) <- colnames(brain.h3k4me3)
```



```{r}

gr.dist = function(gr1, gr2 = NULL, ignore.strand = FALSE,return.matrix=FALSE,as.irange=FALSE, irange.max = 20000, ...)
{
    if (is.null(gr2)){gr2 = gr1}
  
    if(as.irange) {return.matrix=TRUE}

    if (ignore.strand){
        strand(gr1) = '*'
        strand(gr2) = '*'
    }

    ix1 = rep(1:length(gr1), length(gr2))
    ix2 = rep(1:length(gr2), each = length(gr1))

    out = suppressWarnings(distance(gr1[ix1], gr2[ix2], ...))
    
    if(return.matrix){
      out = matrix(out, nrow = length(gr1), ncol = length(gr2))
    }
  return(out)
}
    
    
metagene <- function(gr1,gr2,cutoff = 20000){
  
  ix1 = rep(1:length(gr1), length(gr2))
  ix2 = rep(1:length(gr2), each = length(gr1))
  
  same.chromosome <- as.character(seqnames(gr1[ix1])) == as.character(seqnames(gr2[ix2]))
  minus.strand    <- as.logical(strand(gr2[ix2]) == "-")
  
  distance               <- start(gr1[ix1]) - start(gr2[ix2])
  distance.minus         <- -(end(gr1[ix1]) - end(gr2[ix2]))
  distance[minus.strand] <- distance.minus[minus.strand]
  
  distance[!same.chromosome]       <- NA
  distance[abs(distance) > cutoff] <- NA
  
  to.use     <- which(!is.na(distance))
  
  irange.out <- IRanges(start=distance[to.use],width = width(gr1[ix1])[to.use])
  end(irange.out)[end(irange.out) > cutoff] <- cutoff
  
  irange.out <- c(IRanges(start=-cutoff-1,width=0),irange.out,IRanges(start=cutoff+1,width=0))
  
  return(irange.out)
  }
```





```{r}


tic()
distances.to.promoters.meta <- mclapply(fragments.ls,mc.cores = 4,FUN=function(cell.fragments){
  out <- metagene(cell.fragments,promoters.mOL)
  out
})
toc()

names(distances.to.promoters.meta) <- colnames(brain.h3k4me3)

```

```{r}

cells.depth <- unlist(lapply(fragments.ls,length))

d.coverage <- lapply(seq(distances.to.promoters.meta),function(x){
  norm.factor <- cells.depth[names(distances.to.promoters.meta[x])] /1e3
  x           <- coverage(distances.to.promoters.meta[[x]],shift=20000) /norm.factor
  as.numeric(x)
})

names(d.coverage) <- colnames(brain.h3k4me3)
distances.to.promoters.heatmap <- do.call(rbind,d.coverage)
```



```{r}

spreading.score <- lapply(distances.to.promoters.meta,function(x){
  x <- x[c(-1,-length(x))]
  median(end(x)+start(x))/2
})
spreading.score <- unlist(spreading.score)


mOL.signature$spread.numeric <- spreading.score[rownames(mOL.signature)]
mOL.signature$spread.rank    <- order(mOL.signature$spread.numeric)

ggplot(data=mOL.signature,aes(x=cluster,y=spread.numeric,fill=cluster)) + 
  geom_boxplot(outlier.shape = NA) + 
  coord_flip(ylim = c(-2000,4000)) +
  theme_few() +NoLegend()
  

```

```{r}
spreading.cat             <- Hmisc::cut2(mOL.signature$spread.numeric,g = 10)
names(spreading.cat)      <- rownames(mOL.signature)
mOL.signature$spread.cat  <- spreading.cat[rownames(mOL.signature)]
mOL.signature             <- mOL.signature[!is.na(mOL.signature$spread.cat),]

mOL.signature.cat         <- Hmisc::cut2(mOL.signature$mOL.signature,g = 10)
names(mOL.signature.cat)  <- rownames(mOL.signature)
mOL.signature$mOL.cat     <- mOL.signature.cat[rownames(mOL.signature)]

ylimits=c(0,5e-4)
xlimits=c(-2000,5000)

ggplot(data=mOL.signature[mOL.signature$cluster %in% c("OPC","mOL"),],aes(x=mOL.cat,y=spread.numeric)) + 
  geom_boxplot(outlier.shape = NA) + 
  theme_few() +NoLegend() + 
  scale_fill_manual(values = c("mOL"=scales::hue_pal()(6)[2],"OPC"=scales::hue_pal()(6)[6])) + 
  theme(axis.text.x = element_text(angle=45,hjust=1)) +
  coord_cartesian(ylim=c(-2000,5000))


ggplot(data=mOL.signature[mOL.signature$cluster %in% c("OPC","mOL"),],aes(x=spread.cat,y=mOL.signature,fill=cluster)) + 
  geom_boxplot(outlier.shape = NA) + 
  theme_few() +NoLegend() + 
  scale_fill_manual(values = c("mOL"=scales::hue_pal()(6)[2],"OPC"=scales::hue_pal()(6)[6])) + 
  theme(axis.text.x = element_text(angle=45,hjust=1))

ggplot(data=mOL.signature[sample(rownames(mOL.signature),4000),],aes(x=spread.numeric,y=mOL.signature,col=cluster)) + 
  geom_point(size=0.5) + 
  theme_few() +NoLegend()+ 
  coord_cartesian(xlim=xlimits,ylim=ylimits) + ggtitle("all clusters")

ggplot(data=mOL.signature[!mOL.signature$cluster %in% c("OPC","mOL"),],aes(x=spread.numeric,y=mOL.signature,col=cluster)) + 
  geom_point(size=0.5) + 
  theme_few() +NoLegend()+ 
  coord_cartesian(xlim=xlimits,ylim=ylimits) + ggtitle("Not OLG") #+ scale_color_manual(values = scales::hue_pal()(6)[c(1,3,4,5)])


ggplot(data=mOL.signature[mOL.signature$cluster == "OPC",],aes(x=spread.numeric,y=mOL.signature)) + 
  geom_point(size=0.5,color=scales::hue_pal()(6)[6]) + 
  theme_few() +NoLegend()+ 
  coord_cartesian(xlim=xlimits,ylim=ylimits) + ggtitle("OPCs")


ggplot(data=mOL.signature[mOL.signature$cluster == "mOL",],aes(x=spread.numeric,y=mOL.signature)) + 
  geom_point(size=0.5,col=scales::hue_pal()(6)[2]) + 
  theme_few() +NoLegend()+ 
  coord_cartesian(xlim=xlimits,ylim=ylimits) + ggtitle("mOLs")

```

```{r}
save.image(paste0(params$out_prefix,"/H3K4me3/spreading/spreading.Rdata"))
#load(paste0(params$out_prefix,"/H3K4me3/spreading/spreading.Rdata"))
```


```{r}
# Creat annotation row
ann_row <- mOL.signature[,c('cluster','mOL.signature','spread.numeric'),drop=FALSE]

# Load pseudotime
pt <- readRDS(file=paste0(params$out_prefix,'H3K4me3/pseudotime/pseudotime_OLG_slingshot.Rds'))
pt <- slingPseudotime(pt)[,1,drop=TRUE]
ann_row$pseudotime <- pt[rownames(ann_row)]

########################### OLG lineage

# Sample less cells for faster plotting
OPCs <- rownames(ann_row[ann_row$cluster == 'OPC',])
mOLs <- rownames(ann_row[ann_row$cluster == 'mOL',])

# Select only OPCs and mOLS
ann_row <- ann_row[c(OPCs,mOLs),]
#ann_row <- ann_row[sample(rownames(ann_row),1000),]

ann_row <- ann_row[order(ann_row$mOL.signature),]

# Colors
ann_colors = list('cluster'=CTcolors)


# Adapt the color scale
breaks         <- seq(0,quantile(distances.to.promoters.heatmap,0.99),length.out = 11)

pheatmap(distances.to.promoters.heatmap[rownames(ann_row),17000:30000],
         cluster_cols = FALSE,cluster_rows = FALSE,labels_col = FALSE,labels_row = FALSE,
         annotation_row = ann_row,annotation_colors = ann_colors,
         viridis::viridis(10),breaks=breaks,
         filename = paste0(params$out_prefix,"/H3K4me3/spreading/spreading_OLG_mOLsignature.png"),width = 15,height=15)

```


```{r}
ann_row <- ann_row[order(ann_row$pseudotime),]

pheatmap(distances.to.promoters.heatmap[rownames(ann_row),17000:30000],
         cluster_cols = FALSE,cluster_rows = FALSE,labels_col = FALSE,labels_row = FALSE,
         annotation_row = ann_row,annotation_colors = ann_colors,
         viridis::viridis(10),breaks=breaks,
         filename = paste0(params$out_prefix,"/H3K4me3/spreading/spreading_OLG_pseudotime.png"),width = 15,height=15)


```









```{r}
a   <- Matrix.utils::aggregate.Matrix(x = distances.to.promoters.heatmap[rownames(mOL.signature),],groupings = mOL.signature$cluster,fun='sum')

# Normalize by number of cells
norm.matrix <- table(mOL.signature$cluster)
norm.matrix <- matrix(rep(norm.matrix,dim(a)[1]),nrow=dim(a)[1],ncol=dim(a)[2])
a <- a/norm.matrix

# Change matrix format
a   <- as.data.frame(t(a))
a$x <- 1:dim(a)[1] -20001

# Melt the df for ggplot
a.melted <- reshape2::melt(data = a, id.vars="x")

# Plot
ggplot(data=a.melted) + 
  geom_line(aes(x=x,y=value,col=variable)) + coord_cartesian(xlim=c(-2500,5500)) + 
  theme_few()
```




```{r}
save.image(paste0(params$out_prefix,"/H3K4me3/spreading/spreading.Rdata"))
#load(paste0(params$out_prefix,"/H3K4me3/spreading/spreading.Rdata"))
```





